### PR TITLE
progressive: fix multiple regions in one gfx frame.

### PR DESCRIPTION
**Pull request by llyzs**
[MS-RDPEGFX] 2.2.4.2.1.5 RFX_PROGRESSIVE_REGION

**Note that RFX_PROGRESSIVE_REGION entries that are part of the same frame can share the tiles defined in the tiles field of each entry. In this scenario, tiles are not repeated in successive RFX_PROGRESSIVE_REGION entries.** Across all of the RFX_PROGRESSIVE_REGION entries of a frame, the rectangles (defined in the rects field of each entry) MUST be distinct, and the region defined by these rectangles MUST be completely covered by all of the tiles defined in the RFX_PROGRESSIVE_REGION entries of the frames. **Note that in this context, the frame is bracketed between the RDPGFX_START_FRAME_PDU and the RDPGFX_END_FRAME_PDU, and can span multiple RFX_PROGRESSIVE_FRAME_BEGIN and RFX_PROGRESSIVE_FRAME_END blocks.**

### CODE REVIEW

File: libfreerdp/codec/progressive.c [hunk 1]
```diff
@@ -2383,15 +2398,15 @@ static BOOL progressive_rfx_write_message_progressive_simple(RFX_CONTEXT* contex
...
-	Stream_Write_UINT8(s, msg->numQuant);  /* numQuant (1 byte) */
-	Stream_Write_UINT8(s, 0);              /* numProgQuant (1 byte) */
-	Stream_Write_UINT8(s, 0);              /* flags (1 byte) */
-	Stream_Write_UINT16(s, msg->numTiles); /* numTiles (2 bytes) */
-	Stream_Write_UINT32(s, tilesDataSize); /* tilesDataSize (4 bytes) */
+	Stream_Write_UINT16(s, 0xCCC4);              /* blockType (2 bytes) */
+	Stream_Write_UINT32(s, blockLen);            /* blockLen (4 bytes) */
+	Stream_Write_UINT8(s, 64);                   /* tileSize (1 byte) */
+	Stream_Write_UINT16(s, msg->numRects);       /* numRects (2 bytes) */
+	Stream_Write_UINT8(s, (UINT8)msg->numQuant); /* numQuant (1 byte) */
```
Comment (akallabeth) [inline libfreerdp/codec/progressive.c:2405]: Cast not required

File: libfreerdp/codec/progressive.c [hunk 2]
```diff
@@ -2274,12 +2289,12 @@ INT32 progressive_decompress(PROGRESSIVE_CONTEXT* progressive, const BYTE* pSrcD
 		region16_union_rect(&clippingRects, &clippingRects, &clippingRect);
 	}
 
-	for (i = 0; i < region->numTiles; i++)
+	for (i = 0; i < (UINT16)surface->numUpdatedTiles; i++)
```
Comment (akallabeth) [inline libfreerdp/codec/progressive.c:2292]: Cast not required

File: libfreerdp/codec/progressive.c [hunk 3]
```diff
@@ -2560,7 +2575,7 @@ int progressive_compress(PROGRESSIVE_CONTEXT* progressive, const BYTE* pSrcData,
 	if (!rc)
 		goto fail;
 
-	*pDstSize = Stream_GetPosition(s);
+	*pDstSize = (UINT32)Stream_GetPosition(s);
```
Comment (akallabeth) [inline libfreerdp/codec/progressive.c:2578]: Unrelated change

**Review by akallabeth — APPROVED** [review]
A few minor remarks, other than that LGTM

**Comment by llyzs** [comment]
@akallabeth Those are compiler warnings without cast in MSVC.

**Comment by llyzs** [comment]
OK removed because it is unrelated.

File: libfreerdp/codec/progressive.c [hunk 4]
```diff
@@ -2274,12 +2289,12 @@ INT32 progressive_decompress(PROGRESSIVE_CONTEXT* progressive, const BYTE* pSrcD
 		region16_union_rect(&clippingRects, &clippingRects, &clippingRect);
 	}
 
-	for (i = 0; i < region->numTiles; i++)
+	for (i = 0; i < surface->numUpdatedTiles; i++)
```
Comment (akallabeth) [inline libfreerdp/codec/progressive.c:2292]: type mismatch for index variable, sorry should have pointed that out earlier.

**Review by akallabeth — APPROVED** [review]

**Review by hardening — APPROVED** [review]
LGTM (even if already merged)
